packages/mask/src/plugins/Savings/SNSAdaptor/SavingsForm.tsx

```diff
@@ -0,0 +1,165 @@
+import { EthereumChainBoundary } from '../../../web3/UI/EthereumChainBoundary'
+import { ActionButtonPromise } from '../../../extension/options-page/DashboardComponents/ActionButton'
+
+export interface SavingsFormProps {
+    chainId: number
+    selectedProtocol: number
+    tab: TabType
+    onClose?: () => void
+    onSwapDialogOpen?: () => void
+}
+
+export function SavingsForm({ chainId, selectedProtocol, tab, onClose, onSwapDialogOpen }: SavingsFormProps) {
+    const { t } = useI18N()
+    const { classes } = useStyles()
+    const protocol = SavingsProtocols[selectedProtocol]
+    const targetChainId = chainId
+
+    const web3 = useWeb3(false, targetChainId)
+    const account = useAccount()
+
+    const [inputAmount, setInputAmount] = useState('')
+
+    const { value: nativeTokenBalance, loading: loadingNativeTokenBalance } = useFungibleTokenBalance(
+        EthereumTokenType.Native,
+        '',
+        targetChainId,
+    )
+
+    //#region form controls
+    const formattedBalance = leftShift(Number.parseInt(nativeTokenBalance || '0', 16), 18)
+    const tokenAmount = rightShift(inputAmount || '0', 18)
+    //#endregion
+
+    //#region UI logic
+    // validate form return a message if an error exists
+    const validationMessage = useMemo(() => {
+        if (tokenAmount.isZero()) return t('plugin_trader_error_amount_absence')
+        if (isLessThan(inputAmount, 0)) return t('plugin_trade_error_input_amount_less_minimum_amount')
+
+        const tokenBalance =
+            (tab === TabType.Deposit
+                ? nativeTokenBalance
+                : rightShift(protocol.balance, 18).integerValue().toFixed()) || '0'
+
+        if (isLessThan(tokenBalance, tokenAmount)) {
+            return t('plugin_trader_error_insufficient_balance', {
+                symbol: tab === TabType.Deposit ? protocol.base : protocol.pair,
+            })
+        }
+
+        return ''
+    }, [inputAmount, tokenAmount.toFixed(), formattedBalance])
+    //#endregion
+
+    const needsSwap = protocol.name === 'Lido' && tab === 'withdraw'
+
+    return (
+        <div className={classes.containerWrap}>
+            <Typography variant="h3" className={classes.title}>
+                <img src={IconURLS.lido} className={classes.titleImage} />
+                {tab === TabType.Deposit ? 'Stake' : 'Withdraw'} {protocol.name}
+            </Typography>
+
+            <Typography variant="body1" className={classes.inputLabel}>
+                Balance:{' '}
+                {tab === TabType.Deposit
+                    ? formattedBalance
+                          .toNumber()
+                          .toLocaleString(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 6 })
+                    : protocol.balance}{' '}
+                {tab === TabType.Deposit ? protocol.base : protocol.pair}
+            </Typography>
+
+            <div className={classes.inputWrap}>
+                <TextField
```

Comment:
Refactor to use `TokenAmountPanel` component. It's able to avoid negative numbers and alphabet characters.